pkg/services/publicdashboards/commands/generate_datasources/generate.go

```diff
@@ -0,0 +1,105 @@
+package generate_datasources
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"sync"
+)
+
+type datasourceMeta struct {
+	Slug string `json:"slug"`
+}
+
+type listPluginResponse struct {
+	Items []datasourceMeta `json:"items"`
+}
+
+type datasourceDetails struct {
+	Slug     string `json:"slug"`
+	Backend  bool   `json:"backend"`
+	Alerting bool   `json:"alerting"`
+}
+
+type datasource struct {
+	Json datasourceDetails `json:"json"`
+}
+
+// Get the slugs of all datasource plugins that are compatible with public dashboards
+func GetCompatibleDatasources(baseUrl string) ([]string, error) {
+	slugs, err := getDatasourcePluginSlugs(baseUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	datasources := getDatasourceDetails(slugs, baseUrl)
+
+	// we only consider a datasource to be supported when alerting and backend are both true
+	var supported []string
+	for _, datasource := range datasources {
+		if datasource.Alerting && datasource.Backend {
+			supported = append(supported, datasource.Slug)
+		}
+	}
+
+	sort.Strings(supported)
+	return supported, nil
+}
+
+// Get a list of all the datasource plugin slugs
+func getDatasourcePluginSlugs(baseUrl string) ([]string, error) {
+	resp, err := http.Get(baseUrl + allPluginsEndpoint())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	res := &listPluginResponse{}
+	err = json.NewDecoder(resp.Body).Decode(res)
+	if err != nil {
+		return nil, err
+	}
+	var slugs []string
+	for _, meta := range res.Items {
+		slugs = append(slugs, meta.Slug)
+	}
+	return slugs, nil
+}
+
+// Get the details for each datasource plugin by its slug
+func getDatasourceDetails(slugs []string, baseUrl string) []datasourceDetails {
+	datasources := make([]datasourceDetails, len(slugs))
+	var wg sync.WaitGroup
+	for i, slug := range slugs {
```

Comment:
concurrency!

Reply:
![strongGopher](https://user-images.githubusercontent.com/20195316/198707979-72ef32a7-ddc2-459c-b2fe-4d920776b2dc.png)

Reply:
love this image